File: site/walks.py
"""Exact finite-prefix tools for Basis Walks. Python standard library only.

Constructions: Cambie, Kalviainen and Shallit, Brown-Gerver-Ramsey
Theorems in Small Dimensions, supplied Overleaf snapshot (September 2026).
Finite checks do not prove infinite avoidance. See docs/PROVENANCE.md.
"""
from math import gcd
from functools import reduce

BOUNDS = {3: 7, 4: 4, 5: 4, 6: 3}


def letters(dimension, steps):
    """Generate exactly steps output letters, indexed from zero."""
    if dimension not in BOUNDS or not isinstance(steps, int) or steps < 0:
        raise ValueError("dimension must be 3, 4, 5 or 6; steps a nonnegative integer")
    if dimension == 6:
        # sigma(2n)=-sigma(n), sigma(2n+1)=1-sigma(n), modulo 4.
        states = [0] * (steps + 1)
        for n in range(1, steps + 1):
            states[n] = ((n % 2) - states[n // 2]) % 4
        coding = {(0, 1): 0, (1, 2): 1, (2, 3): 2, (3, 0): 3,
                  (0, 2): 4, (1, 3): 4, (2, 0): 5, (3, 1): 5}
        return [coding[states[n], states[n + 1]] for n in range(steps)]
    # Return-word construction, expanded directly from the five-letter morphism.
    tau = dict(zip("ABCDE", ("AB", "AACA", "ADE", "AACCE", "ADCCA")))
    code = dict(zip("ABCDE", ("0", "1", "20", "21", "21") if dimension == 3
                    else ("0", "1", "2", "3", "3")))
    word = "A"
    while len(word) < steps:
        word = "".join(tau[a] for a in word)
    return [int(a) for a in "".join(code[a] for a in word)[:steps]]


def vertices(dimension, steps):
    points = [[0] * dimension]
    for a in letters(dimension, steps):
        point = points[-1].copy()
        point[a] += 1
        points.append(point)
    return points


def anchor_check(points, anchor, forbidden):
    """All lines from one anchor to later points, exact primitive directions.

    A forbidden collinear set has an earliest vertex, so scanning every anchor
    covers unequal spacings too. Returns indices of one witness, or None.
    """
    directions = {}
    for j in range(anchor + 1, len(points)):
        delta = [b - a for a, b in zip(points[anchor], points[j])]
        divisor = reduce(gcd, delta)
        if not divisor:
            raise ValueError("duplicate vertices are not supported")
        key = tuple(x // divisor for x in delta)
        group = directions.setdefault(key, [])
        group.append(j)
        if len(group) >= forbidden - 1:
            return [anchor] + group[:forbidden - 1]
    return None


def check(dimension=6, steps=128):
    """Small bounded browser/example check; use verify.py for resumable runs."""
    if not 0 <= steps <= 512:
        raise ValueError("Interactive check limited to 512 steps; use verify.py for larger runs")
    points = vertices(dimension, steps)
    for anchor in range(len(points)):
        witness = anchor_check(points, anchor, BOUNDS[dimension])
        if witness:
            return {"dimension": dimension, "steps": steps, "forbidden": BOUNDS[dimension],
                    "status": "counterexample", "witness": witness}
    return {"dimension": dimension, "steps": steps, "vertices": len(points),
            "forbidden": BOUNDS[dimension], "status": "finite-prefix-pass",
            "note": "Not an infinite proof"}

File: site/walks.js
// Integer construction logic, deliberately independent of Python's 3D/4D morphism.
export const bounds = {3: 7, 4: 4, 5: 4, 6: 3};
export const colors = ['#3183df', '#d58727', '#179b80', '#be669c', '#8263ce', '#dc6659'];
export function state(n, base) {
  let even = 0, odd = 0, position = 0;
  while (n) {
    const digit = n % base;
    if (position++ % 2) odd += digit; else even += digit;
    n = Math.floor(n / base);
  }
  return base === 2 ? ((even - odd) % 4 + 4) % 4 : [even % 2, odd % 2];
}
export function letters(d, n) {
  if (!(d in bounds) || !Number.isInteger(n) || n < 0 || n > 100000) throw Error('Invalid dimension or prefix length');
  if (d === 6) {
    const code = {'01':0, '12':1, '23':2, '30':3, '02':4, '13':4, '20':5, '31':5};
    return Array.from({length:n}, (_, j) => code[`${state(j,2)}${state(j+1,2)}`]);
  }
  const output = []; let previous = 0;
  for (let j = 1; output.length < n; j++) {
    const [a,b] = state(j,3);
    if (a || b) continue;
    const r = (j-previous)/2; previous = j;
    if (d === 3) output.push(...[[0], [1], [2,0], [2,1]][r-1]);
    else output.push(r-1);
  }
  return output.slice(0,n);
}
export function vertices(d, n) {
  const points = [Array(d).fill(0)];
  for (const letter of letters(d,n)) {
    const p = [...points.at(-1)]; p[letter]++; points.push(p);
  }
  return points;
}
const gcd = (a,b) => b ? gcd(b,a%b) : a;
export function check(d,n) {
  if (n > 512) throw Error('Interactive checking is limited to 512 steps');
  const points = vertices(d,n); let pairs = 0;
  for (let a=0; a<points.length; a++) {
    const directions = new Map();
    for (let b=a+1; b<points.length; b++) {
      const delta = points[b].map((x,k) => x-points[a][k]);
      const divisor = delta.reduce(gcd,0);
      const key = delta.map(x => x/divisor).join(',');
      const group = directions.get(key) || [];
      group.push(b); directions.set(key,group); pairs++;
      if (group.length >= bounds[d]-1) return {status:'counterexample', witness:[a,...group], pairs};
    }
  }
  return {status:'finite-prefix-pass', dimension:d, steps:n, vertices:n+1, forbidden:bounds[d], pairs};
}
// 4D/5D: the paper's return-displacement map. 6D: an illustrative
// real-linear map, NOT an integer projection certified for the infinite walk.
export function shadow(p) {
  if (p.length === 3) return [...p];
  const columns = p.length < 6 ? [[1,0,1],[-1,1,2],[0,-1,3],[-2,0,4],[0,0,0]] :
    [[1,0,0.7],[-0.5,0.866,-0.7],[-0.5,-0.866,0.3],[0.866,0.5,-0.3],[-0.866,0.5,-1],[0,-1,1]];
  return [0,1,2].map(k => p.reduce((sum,x,j) => sum+x*columns[j][k],0));
}
